File: ASD Prototype/Utils/LAP/RLAP.cpp
//
//  lapjv.cpp
//  ASD Prototype
//
//  Created by Benjamin Lee on 6/17/25.
//

#include <memory>
#include <cmath>
#include "RLAP.h"
#include "CrouseSAP.h"


int solve_rlapf(int numrows, int numcols,
                const float* cost, bool maximize,
                int* rowsol, int* colsol) noexcept {
    return solve_rectangular_linear_assignment<int, float>(numrows,
                                                           numcols,
                                                           cost,
                                                           maximize,
                                                           rowsol,
                                                           colsol);
}

int solve_rlap(int numrows, int numcols,
                const double* cost, bool maximize,
                int* rowsol, int* colsol) noexcept {
    return solve_rectangular_linear_assignment<int, double>(numrows,
                                                            numcols,
                                                            cost,
                                                            maximize,
                                                            rowsol,
                                                            colsol);
}

File: ASD Prototype/ASD/Video/Tracking/LAP/RLAP.swift
//
//  LAP.swift
//  ASD Prototype
//
//  Created by Benjamin Lee on 6/17/25.
//

import Foundation
import RLAP

extension ASD.Tracking {
    @usableFromInline
    @discardableResult
    static func solveRLAP(
        dims: (Int, Int),
        cost: [Float],
        rows: inout [Int],
        cols: inout [Int],
        maximize: Bool = false
    ) -> Int {
        rows = [Int](repeating: 0, count: dims.0)
        cols = [Int](repeating: 0, count: dims.1)
        
        // Some very safe code the definitely won't cause segfaults :)
        return solve_rlapf(
            dims.0,
            dims.1,
            cost.withUnsafeBufferPointer { $0.baseAddress! },
            maximize,
            rows.withUnsafeMutableBufferPointer { $0.baseAddress! },
            cols.withUnsafeMutableBufferPointer { $0.baseAddress! }
        )
    }
    
    @usableFromInline
    @discardableResult
    static func solveRLAP(
        dims: (Int, Int),
        cost: [Double],
        rows: inout [Int],
        cols: inout [Int],
        maximize: Bool = false
    ) -> Int {
        rows = [Int](repeating: 0, count: dims.0)
        cols = [Int](repeating: 0, count: dims.1)
        
        // Some very safe code the definitely won't cause segfaults :)
        return solve_rlap(
            dims.0,
            dims.1,
            cost.withUnsafeBufferPointer { $0.baseAddress! },
            maximize,
            rows.withUnsafeMutableBufferPointer { $0.baseAddress! },
            cols.withUnsafeMutableBufferPointer { $0.baseAddress! }
        )
    }
}
